pkg/user/external_server.go

```diff
@@ -0,0 +1,227 @@
+	if err := json.NewDecoder(req.Body).Decode(body); err != nil {
+		l.WithError(err).Error("[ExternalUserManagerServer] Bad request")
+		http.Error(rw, err.Error(), http.StatusBadRequest)
+		return false
+	}
+	return true
+}
+
+func sendResponse(rw http.ResponseWriter, status int, body interface{}) {
+	rw.Header().Set("Content-Type", "application/json")
+	rw.WriteHeader(status)
+
+	if body != nil {
+		// Set up the pipe to write data directly into the Reader.
+		pr, pw := io.Pipe()
+
+		// Write JSON-encoded data to the Writer end of the pipe.
+		// Write in a separate concurrent goroutine, and remember
+		// to Close the PipeWriter, to signal to the paired PipeReader
+		// that we’re done writing.
+		go func() {
+			pw.CloseWithError(json.NewEncoder(pw).Encode(body))
+		}()
+
+		_, err := io.Copy(rw, pr)
+		if err != nil {
+			log.WithError(err).Warn("[ExternalServer] error writing response")
+		}
+	}
+}
+
+func StartExternalServer(listenAddr string, stop <-chan struct{}, um UserManager) error {
+	ctx := context.Background()
+
+	handler := func(rw http.ResponseWriter, req *http.Request) {
+		l := log.WithField("path", req.URL.Path).WithField("method", req.Method)
+		l.Debug("[ExternalUserManagerServer] request received")
+		defer func() {
+			if r := recover(); r != nil {
+				l.WithField("panic", r).Error("[ExternalUserManagerServer] panic recovered")
+				http.Error(rw, fmt.Sprintf("%v", r), http.StatusInternalServerError)
+			}
+		}()
+		switch fmt.Sprintf("%s %s", req.Method, req.URL.Path) {
+		case "PUT /user/admin":
+			var u User
+			if readRequestBody(l, rw, req, &u) {
+				err := um.NewAdmin(&u)
+				if err != nil {
+					l.WithError(err).Error("[ExternalUserManagerServer] Admin user creation failure")
+					http.Error(rw, err.Error(), http.StatusInternalServerError)
+					return
+				}
+				sendResponse(rw, http.StatusOK, nil)
+			}
+		case "PUT /user":
+			var u NewUserRequest
+			if readRequestBody(l, rw, req, &u) {
+				nu, err := um.New(u.Name, u.Email, u.Password)
+				if err != nil {
+					l.WithError(err).Error("[ExternalUserManagerServer] Admin user creation failure")
+					http.Error(rw, err.Error(), http.StatusInternalServerError)
+					return
+				}
+				sendResponse(rw, http.StatusOK, nu)
+			}
+		case "GET /user":
+			var q Query
+			if readRequestBody(l, rw, req, &q) {
+				u, err := um.Get(&q)
+				if err != nil {
+					l.WithError(err).Error("[ExternalUserManagerServer] User query")
+					http.Error(rw, err.Error(), http.StatusInternalServerError)
+					return
+				}
+				sendResponse(rw, http.StatusOK, u)
+			}
+		case "POST /user":
+			var u User
+			if readRequestBody(l, rw, req, &u) {
+				nu, err := um.Update(&u)
+				if err != nil {
+					l.WithError(err).Error("[ExternalUserManagerServer] User update failure")
+					http.Error(rw, err.Error(), http.StatusInternalServerError)
+					return
+				}
+				sendResponse(rw, http.StatusOK, nu)
+			}
+		case "PUT /user/import":
```

Comment:
what is it actually importing? gateway importing gateway users via dotmesh?

Reply:
It's like creating a user, except that it's passed a ready-made user object that just needs storing - the difference being that normal user creation hashes and salts the password. I found it used by the "restore etcd from a backup" thing inside dotmesh, and it's also used by the gateway (which imports the dotmesh user package directly) in `createAccountInEtcd`.